Add Application insights to React App

mhassaninmsft opened the pull request:
1. Updates the Front end application to use Application insights using the MSFT react plugin. []
2. Updates instructions for creation of dev environments to make it safer and easier to update application insights `ikey` and signalR service backend.

Commit:
Updating packages

mhassaninmsft committed:
updates

Commit:
adding initial telemetry client implementation

Commit:
removing telemetry

mhassaninmsft committed:
react plugin

Commit:
Merge branch 'dev' of https://github.com/NarmathaBala/vsts-extension-retrospectives into nabalas/update-telemetry

mhassaninmsft committed:
app insights all components

mhassaninmsft committed:
telemetry updates

mhassaninmsft commented on RetrospectiveExtension.Backend/appsettings.json:
```diff
@@ -11,5 +11,8 @@
         "Default": "Warning"
       }
     }
+  },
```
Adding Default placeholder in appsettings for the backend to make it for local debugging/testing to change this

mhassaninmsft committed:
fix merge conflicts

mindlessroman commented on RetrospectiveExtension.Frontend/.gitignore:
```diff
@@ -0,0 +1,2 @@
+# Development version to publish dev version of extension
+vss-extension-dev.json
```
do we need this nested `.gitignore`? If so, the commented out one should be removed from the main one. But i'm not sure we need a separate one for the frontend.

mhassaninmsft replied:
will leave everything in the top level .gitignore since many folks left similar comments

polatengin commented on CONTRIBUTING.md:
```diff
@@ -38,11 +38,20 @@ Clone the repository to your local machine from the Azure DevOps endpoint.
 - To test your changes, you will need to publish a new extension under a new Azure DevOps publisher account. Refer to the [documentation](https://docs.microsoft.com/en-us/azure/devops/extend/publish/overview?view=vsts) on publishing extensions. You can publish it to any test Azure DevOps organization that you are an admin of (As a Microsoft employee, you can create a new test organization from yo ...
 
-- Update the 'vss-extension-dev.json' file with the new publisher that you setup. Also update the name and id fields.
+- Copy the file `vss-extensions-dev.json.template` into `vss-extensions-dev.json` file with the new publisher that you setup. Also update the name and id fields.
```
`vss-extension-dev.json` (singular) file ignored in the `.gitignore` file, here we're stating the file as plural `(vss-extensions-dev.json)`

either of them should be changed I think.

mhassaninmsft replied:
changed file name to `vss-extension-dev` singular form

polatengin commented on RetrospectiveExtension.Frontend/components/feedbackCarousel.tsx:
```diff
@@ -24,7 +26,7 @@ export default class FeedbackCarousel extends React.Component<IFeedbackCarouselP
       .map((columnItem) => {
         const feedbackItemProps =
-          FeedbackColumn.createFeedbackItemProps(feedbackColumnProps, columnItem, true);
+        FeedbackColumnStatic.createFeedbackItemProps(feedbackColumnProps, columnItem, true);
```
I'd move `createFeedbackItemProps` static function to a utility class and use it, instead of exporting same class with two different approaches (withAITracking and with static)

it seems `createFeedbackItemProps` method is the only method we need here

mhassaninmsft replied:
good point. I will create a utility class and name `FeedbackColumnHelperClass`

mindlessroman commented on RetrospectiveExtension.Frontend/config/.gitignore:
```diff
@@ -0,0 +1,2 @@
+# ignore enviornment containing possibly sensitive information such as the backend/ instrumentation key
+environment.tsx
```
Or this separate `.gitignore` - it makes more sense to me to just have the toplevel one and specify all the files that may have sensitive keys in them.

polatengin commented on RetrospectiveExtension.Frontend/components/feedbackColumn.tsx:
```diff
@@ -300,3 +302,7 @@ export default class FeedbackColumn extends React.Component<FeedbackColumnProps,
   }
 }
+
+export { FeedbackColumn as FeedbackColumnStatic}; // static functions are not preserved with `withAITracking`,
```
can we move static methods to a utility class and not export the same class in a different approach?

mhassaninmsft replied:
makes sense

polatengin commented on RetrospectiveExtension.Frontend/components/feedbackItem.tsx:
```diff
@@ -934,7 +936,7 @@ export default class FeedbackItem extends React.Component<IFeedbackItemProps, IF
               {this.state.searchedFeedbackItems
                 .map((columnItem) => {
-                  const feedbackItemProps = FeedbackColumn.createFeedbackItemProps(
+                  const feedbackItemProps = FeedbackColumnStatic.createFeedbackItemProps(
```
same with the previous comment

mhassaninmsft replied:
done, created `FeedbackColumnHelper` class

mindlessroman commented on RetrospectiveExtension.Frontend/config/environment.tsx.template:
```diff
@@ -0,0 +1,7 @@
+const environment = {
```
I appreciate this being template-able! 👏

phongcao replied:
In the docs, we use the following default values:

```json
{
   "CollaborationStateServiceUrl": "https://my-backend-service.com", // change this to the deployed backend service
   "AppInsightsInstrumentKey": "my_instrumentation_key" // put Instrumentation key here
}
```

Can you either update the docs or here for consistency?

mhassaninmsft replied:
done

polatengin commented on RetrospectiveExtension.Frontend/components/feedbackItemGroup.tsx:
```diff
@@ -49,7 +51,7 @@ export default class FeedbackItemGroup extends React.Component<IFeedbackItemGrou
     // const droppedItemId = e.dataTransfer.getData('id');
     const droppedItemId = localStorageHelper.getIdValue();
-    FeedbackItem.handleDropFeedbackItemOnFeedbackItem(this.props.mainFeedbackItem, droppedItemId, this.props.mainFeedbackItem.id);
+    FeedbackItemStatic.handleDropFeedbackItemOnFeedbackItem(this.props.mainFeedbackItem, droppedItemId, this.props.mainFeedbackItem.id);
```
same with the previous comment

_this time for `handleDropFeedbackItemOnFeedbackItem` method_

mindlessroman commented on RetrospectiveExtension.Frontend/index.tsx:
```diff
@@ -6,13 +6,15 @@ import { init as sdkInit }  from 'azure-devops-extension-sdk';
 
 import FeedbackBoardContainer, { FeedbackBoardContainerProps } from './components/feedbackBoardContainer';
-// TODO (enpolat) : import { appInsightsClient, TelemetryEvents } from './utilities/appInsightsClient'
+// import { TelemetryClient } from './utilities/external/telemetryclient';
```
We shouldn't be committing commented out code

mindlessroman commented on RetrospectiveExtension.Frontend/index.tsx:
```diff
@@ -6,13 +6,15 @@ import { init as sdkInit }  from 'azure-devops-extension-sdk';
-// TODO (enpolat) : appInsightsClient.trackEvent(TelemetryEvents.ExtensionLaunched);
+appInsights.trackEvent({name:"Startup",properties:{"version":"2"}});
+appInsights.trackMetric({name:"startup Time",average:5});
+// let client = TelemetryClient.getClient();
```
Mindful of commented out code.

phongcao replied:
nit: missing space after `,`

phongcao replied:
I don't really understand about those properties (`version`, `average`...) so can you please explain more about them in our standup?

NarmathaBala replied:
Re:tracking versions - It is a great idea to track version information. AppInsights provides an option to set context with addTelemetryInitializers, that way the context of, say a user or version data is tracked consistently throughout. 
Re: tracking startup time - The trackMetric looks like a test code that need to be removed

phongcao replied:
Thanks, it makes sense to me that we should remove the trackMetric code. For tracking version, shouldn't we pull the `version` value from some configuration (for example app insights) object? Also, the `Startup` custom event with `version` property seems confusing to me.

polatengin reviewed (approved):
I left a few comments, but so far it looks very good. if you can address those comments, it'd be even better. thanks 🥳👍

mindlessroman commented on RetrospectiveExtension.Frontend/components/workflowStage.tsx:
```diff
@@ -1,6 +1,8 @@
 import * as React from 'react';
 import { WorkflowPhase } from '../interfaces/workItem';
+import { withAITracking } from '@microsoft/applicationinsights-react-js';
+import { reactPlugin, appInsights } from '../utilities/external/telemetryClient';
```
I'd assume this would fail if we're removing that elsewhere in the PR?

mindlessroman replied:
NEVERMIND.... I didn't see the capitalization change

mindlessroman commented on RetrospectiveExtension.Frontend/components/noFeedbackBoardsView.tsx:
```diff
@@ -24,3 +26,6 @@ export default class NoFeedbackBoardsView extends React.Component<NoFeedbackBoar
     )
   }
 }
+
```
Probably only need one new line here

mindlessroman replied:
(and elsewhere)

mindlessroman commented on RetrospectiveExtension.Backend/appsettings.json:
```diff
@@ -11,5 +11,8 @@
       }
     }
+  },
+  "ApplicationInsights":{
```
Just for uniformity's sake - probably could have spaces between the colons and the subsequent characters

mindlessroman commented on RetrospectiveExtension.Frontend/components/selectorCombo.tsx:
```diff
@@ -300,3 +302,6 @@ export default class SelectorCombo<T> extends React.Component<ISelectorComboProp
 }
+
+export default  SelectorCombo;
+// export default withAITracking(reactPlugin,SelectorCombo); //mhassani: TODO: Figure out how to make App insights export generic classes
```
No lingering TODOs before merging. (mentioning here to keep track of it)

phongcao replied:
Agree, we should remove line 307 to avoid confusion.

mindlessroman commented on .gitignore:
```diff
@@ -266,3 +266,5 @@ RetrospectiveExtension.FrontEnd/vss-extension.dev.json
 .env
 allowed_origins.json
 dev_certs.json
+
```
We should include anything we provide templates for here.

phongcao replied:
Similar to Hannah's concern, can we remove line 270 to make the file look cleaner?

mhassaninmsft replied:
done added templates in the same place for consistency

mindlessroman reviewed (commented):
I had a general question about the naming convention of "withAITracking" - would love to know more.

phongcao commented on RetrospectiveExtension.Frontend/components/actionItemDisplay.tsx:
```diff
@@ -316,3 +318,4 @@ export default class ActionItemDisplay extends React.Component<ActionItemDisplay
     );
   }
 }
+export default withAITracking(reactPlugin,ActionItemDisplay);
```
nit: Add a space after `,`. Please fix this issue in other places as well.

phongcao commented on RetrospectiveExtension.Frontend/components/effectivenessMeasurementRow.tsx:
```diff
@@ -20,8 +22,7 @@ export interface EffectivenessMeasurementRowProps {
 }
-
-export default class EffectivenessMeasurementRow extends React.Component<EffectivenessMeasurementRowProps, EffectivenessMeasurementRowState> {
+ class EffectivenessMeasurementRow extends React.Component<EffectivenessMeasurementRowProps, EffectivenessMeasurementRowState> {
```
nit: add empty line above. Same for other places.

phongcao commented on RetrospectiveExtension.Frontend/components/feedbackCarousel.tsx:
```diff
@@ -77,7 +79,7 @@ export default class FeedbackCarousel extends React.Component<IFeedbackCarouselP
               // @ts-ignore TS2786
               <Slider {...settings}>
-                {React.Children.map(this.renderFeedbackCarouselItems(columnProps), (child: React.ReactElement<FeedbackItem>) => {
+                {React.Children.map(this.renderFeedbackCarouselItems(columnProps), (child: React.ReactElement<typeof FeedbackItem>) => {
```
Can you explain why we need to add `typeof` here?

mhassaninmsft replied:
That is a good question. I guess it is because `withAITracking` takes a class as input and returns a value instead of another class. Hence if we remove the `typeof`, the compiler would complain that a value is being used a class. In Typescript there are little difference between an actual class and an object that has the same fields/parameters as defined in a class. Expect in small ways such as typeof. The original function `React.ReactElement<T>` expected either a class or a string,
`interface ReactElement<P = any, T extends string | JSXElementConstructor<any> = string | JSXElementConstructor<any>>`  I am not sure why the function `withAITracking` in the app insights SDK does not return an object that implments `JSXElementConstructor` otherwise the `typeof` expression would have been needless. I can look deeper into this if needed.

phongcao replied:
No need, thanks for your explanation!

phongcao commented on RetrospectiveExtension.Frontend/components/feedbackItem.tsx:
```diff
@@ -977,3 +979,7 @@ export default class FeedbackItem extends React.Component<IFeedbackItemProps, IF
 }
+
+export {FeedbackItem as FeedbackItemStatic }; // static functions are not preserved with `withAITracking`,
+// Hence to use any static function the untransofrmed class has to exported first
```
nit: can we fix the typo in this comment and re-format it so that it's easier to read?

mhassaninmsft replied:
removed comments since it is not needed any more

phongcao commented on RetrospectiveExtension.Frontend/components/feedbackItemGroup.tsx:
```diff
@@ -109,3 +111,6 @@ export default class FeedbackItemGroup extends React.Component<IFeedbackItemGrou
   }
 }
+
+
```
nit: redundant line

phongcao commented on RetrospectiveExtension.Frontend/index.tsx:
```diff
@@ -6,13 +6,15 @@ import { init as sdkInit }  from 'azure-devops-extension-sdk';
 import { isHostedAzureDevOps } from './utilities/azureDevOpsContextHelper';
 import { getProjectId } from './utilities/servicesHelper';
 import './css/main.scss';
+import {  appInsights } from './utilities/external/telemetryClient';
```
nit: There seems to be a redundant space after `{`

phongcao commented on RetrospectiveExtension.Frontend/utilities/external/telemetryClient.ts:
```diff
@@ -0,0 +1,17 @@
+    }
+});
+appInsights.loadAppInsights();
+export { reactPlugin, appInsights };
```
nit: we should have an empty line above

phongcao replied:
nit: this file should use 2-space indentation like other files for consistency.

mhassaninmsft replied:
The `trackEvent` and `trackMetric` events were for local debugging and testing which I forgot to remove. I removed those lines now

phongcao replied:
Thanks for removing those lines. Can you also help format the file to use 2-space indent if it hasn't already used it?

mhassaninmsft replied:
used 2 line indents in file formatting options

phongcao commented on RetrospectiveExtension.Frontend/utilities/external/telemetryClient.ts:
```diff
@@ -0,0 +1,17 @@
+import { ApplicationInsights, SeverityLevel } from "@microsoft/applicationinsights-web"
+import environment from '../../config/environment'
+
+
```
nit: redundant line here

phongcao commented on RetrospectiveExtension.Frontend/utilities/external/telemetryClient.ts:
```diff
@@ -0,0 +1,17 @@
+import { ReactPlugin } from '@microsoft/applicationinsights-react-js';
+import { ApplicationInsights, SeverityLevel } from "@microsoft/applicationinsights-web"
```
nit: use single quote

phongcao commented on RetrospectiveExtension.Frontend/utilities/external/telemetryClient.ts:
```diff
@@ -0,0 +1,17 @@
+const reactPlugin = new ReactPlugin();
+const appInsights = new ApplicationInsights({
+    config: {
+        instrumentationKey: environment.AppInsightsInstrumentKey,
```
Do you know what happens if the `instrumentationKey` is empty?

mhassaninmsft replied:
no crashes happen and the application silently continues. If app insights is required to start the App, we can add a check to make sure that it is not empty and perhaps issue an error or warning if it is.

phongcao replied:
That's good enough for now, thanks!

phongcao commented:
Thanks for adding this feature. Most of my comments are related to code formatting. However, I do have one question about the event/metric that we're tracking so let's chat about that more in our stand up.

mhassaninmsft commented:
> 

so withAITracking is "with Application Insights Tracking". I guess the authors of the library could have done a better job with cases I guess

phongcao commented:
@mhassaninmsft @NarmathaBala Sorry if this is something that I missed in our last demo. Can we pull something like the following picture in our environment to see the view counts of individual components?

![image](https://user-images.githubusercontent.com/9741323/151207253-8644f236-e309-4aaf-80d8-0d56b5ddc0bf.png)

Commit:
Merge branch 'dev' of https://github.com/NarmathaBala/vsts-extension-retrospectives into nabalas/update-telemetry

mhassaninmsft commented:
> @mhassaninmsft @NarmathaBala Sorry if this is something that I missed in our last demo. Can we pull something like the following picture in our environment to see the view counts of individual components?
> 
> ![image](https://user-images.githubusercontent.com/9741323/151207253-8644f236-e309-4aaf-80d8-0d56b5ddc0bf.png)

@phongcao yes that is what we will be adding to the dashboards. If there are any more insights metrics that you think would be helpful, I would love to pair with you on this task if you have some time

mhassaninmsft committed:
PR changes 1

mhassaninmsft committed:
Merge branch 'nabalas/update-telemetry' of github.com:NarmathaBala/vsts-extension-retrospectives into nabalas/update-telemetry

mhassaninmsft committed:
PR review 2

mhassaninmsft committed:
test commit

mhassaninmsft committed:
test commit

mhassaninmsft committed:
test commit 2

mhassaninmsft committed:
update git ignore

mhassaninmsft committed:
update git ignore

mhassaninmsft commented on .gitignore:
```diff
@@ -260,9 +260,9 @@ paket-files/
 .DS_Store
 
 # dev files
-RetrospectiveExtension.FrontEnd/vss-extension.dev.json
```
When the folder contains a `.` gitignore does not honor the file name. that is why I created a separate .gitignore file inside `RetrospectiveExtension.FrontEnd`

mhassaninmsft commented:
addressed the PR reviews feedback. There might have been some formatting and linting issues that I missed. But with Phong's PR for auto Lints this should not be a problem

phongcao commented:
@mhassaninmsft Can you resolve conflicts and update your branch with `dev`?

mhassaninmsft committed:
merge with dev

mhassaninmsft committed:
add enviornment.tsx

mhassaninmsft committed:
add enviornment.tsx

mhassaninmsft commented on RetrospectiveExtension.Frontend/config/environment.tsx:
```diff
@@ -1,7 +1,7 @@
 const environment = {
-  CollaborationStateServiceUrl: '<collaboration state service url>',
+  CollaborationStateServiceUrl: 'https://my-backend-service.com', // change this to the deployed backend service
```
Although `enviornment.tsx`  is gitignored, the file is still needed by the code to compile and the CI/CD tests depend on it. So I left the file as is, but it is still gitignored.

mhassaninmsft commented:
> 

@phongcao done, resolved and rebased with `dev` branch. Thanks!

phongcao commented on RetrospectiveExtension.Frontend/components/feedbackBoardPreviewEmail.tsx:
```diff
@@ -16,8 +18,7 @@ interface IFeedbackBoardPreviewEmailProps {
 }
-
-export default class FeedbackBoardPreviewEmail extends React.Component<IFeedbackBoardPreviewEmailProps, IFeedbackBoardPreviewEmailState> {
+class FeedbackBoardPreviewEmail extends React.Component<IFeedbackBoardPreviewEmailProps, IFeedbackBoardPreviewEmailState> {
```
nit: missing an empty line above.

phongcao commented on RetrospectiveExtension.Frontend/components/feedbackColumn.tsx:
```diff
@@ -300,3 +258,52 @@ export default class FeedbackColumn extends React.Component<FeedbackColumnProps,
+  }
+}
+
+export default  withAITracking(reactPlugin, FeedbackColumn);
```
nit: redundant space after default

phongcao commented on RetrospectiveExtension.Frontend/components/feedbackColumn.tsx:
```diff
@@ -143,51 +145,7 @@ export default class FeedbackColumn extends React.Component<FeedbackColumnProps,
-  }
-
-  private renderFeedbackItems = () => {
+  protected renderFeedbackItems = () => {
```
Seems to me that this doesn't need to be changed to `protected` right?

mhassaninmsft replied:
yeah, changed it back to private. It was some experiments I was doing with the static helper class that I created, but forgot to revert it back to private.

mhassaninmsft committed:
cleanup 1

phongcao commented on RetrospectiveExtension.Frontend/components/feedbackItem.tsx:
```diff
@@ -977,3 +963,26 @@ export default class FeedbackItem extends React.Component<IFeedbackItemProps, IF
+export class FeedbackItemHelper
+{
+// Handle linking/grouping workitems and reload any updated items.
+public static handleDropFeedbackItemOnFeedbackItem = async (feedbackItemProps: IFeedbackItemProps, droppedItemId: string, targetItemId: string) => {
```
nit: something wrong with the indentation for this method

mhassaninmsft replied:
fixed indentation

phongcao commented on RetrospectiveExtension.Frontend/index.tsx:
```diff
@@ -6,13 +6,9 @@ import { init as sdkInit }  from 'azure-devops-extension-sdk';
 import './css/main.scss';
+import { appInsights } from './utilities/external/telemetryClient';
 
 import FeedbackBoardContainer, { FeedbackBoardContainerProps } from './components/feedbackBoardContainer';
```
Why did you remove `initializeIcons()` method?

mhassaninmsft replied:
added it back, it was between two lines of TODO and must have been deleted by mistake

mhassaninmsft replied:
good catch! :)

mhassaninmsft committed:
cleanup 2

phongcao commented:
Thanks for addressing all feedback, great work! @mhassaninmsft @NarmathaBala :+1:

mhassaninmsft commented:
> Thanks for addressing all feedback, great work! @mhassaninmsft @NarmathaBala

Thank you Phong for the quick and thorough review. Really appreciate it
